### PR TITLE
basic_parser rewrite: header-only

### PR DESCRIPTION
**Pull request by vinniefalco**
The basic_parser is rewritten to be header-only. The nodejs parser is removed from the include subtree and placed into the test directory. Other changes:

* Parser specific error codes in parse_error.hpp
* Add parser-bench performance testing, nodejs vs beast
* New random message generator for fuzz tests
* Test for header-only parser using random message generator
* Augmented some existing message tests to check more cases
* Add beast::write(): This function performs serialization of a variable list of arguments to a streambuf. It accepts a wide variety of argument types, using boost::asio::buffer and boost::lexical_cast where possible.

First 2 commits already passed review (add WebSocket tests and Beast documentation work). Preview of the documentation changes are here:
http://vinniefalco.github.io/stage/beast

This commit also already passed review (Add UTF8_checker unit test)

Reviewers: @nbougalis

### CODE REVIEW

File: examples/sig_wait.h [hunk 1]
```diff
@@ -32,18 +32,11 @@ sig_wait()
     boost::asio::io_service ios;
     boost::asio::signal_set signals(
         ios, SIGINT, SIGTERM);
-    std::mutex m;
-    bool stop = false;
-    std::condition_variable cv;
     signals.async_wait(
-        [&](boost::system::error_code const&, int)
+        [&](boost::system::error_code const&, int sig)
         {
```
Comment (miguelportilla) [inline examples/sig_wait.h]: Unused local variable 'sig'.

File: include/beast/http/parser.hpp [hunk 2]
```diff
@@ -57,94 +65,164 @@ class parser
...
+            [&](char const* s, method_t m)
+            {
+                if(this->method_ == s)
+                {
+                    m_.method = m;
+                    return true;
+                }
+                return false;
+            };
+        while(false)
```
Comment (miguelportilla) [inline include/beast/http/parser.hpp]: Non-executing code.

File: include/beast/http/detail/basic_parser.hpp [hunk 3]
```diff
@@ -0,0 +1,388 @@
...
+        1, 1, 1, 1,  1, 1, 1, 1,  1, 1, 1, 1,  1, 1, 1, 1, // 128
+        1, 1, 1, 1,  1, 1, 1, 1,  1, 1, 1, 1,  1, 1, 1, 1, // 144
+        1, 1, 1, 1,  1, 1, 1, 1,  1, 1, 1, 1,  1, 1, 1, 1, // 160
+        1, 1, 1, 1,  1, 1, 1, 1,  1, 1, 1, 1,  1, 1, 1, 1, // 176
+        1, 1, 1, 1,  1, 1, 1, 1,  1, 1, 1, 1,  1, 1, 1, 1, // 192
+        1, 1, 1, 1,  1, 1, 1, 1,  1, 1, 1, 1,  1, 1, 1, 1, // 208
+        1, 1, 1, 1,  1, 1, 1, 1,  1, 1, 1, 1,  1, 1, 1, 1, // 224
+        1, 1, 1, 1,  1, 1, 1, 1,  1, 1, 1, 1,  1, 1, 1, 1  // 240
+    }};
+    return tab[static_cast<std::uint8_t>(c)] != 0;
```
Comment (miguelportilla) [inline include/beast/http/detail/basic_parser.hpp]: A bitset may be cache friendlier.

     static std::bitset<256> tab {
        "11111111111111111111111111111111"
        "11111111111111111111111111111111"
        "11111111111111111111111111111111"
        "11111111111111111111111111111111"
        "01111111111111111111111111111111"
        "11111111111111111111111111111111"
        "11111111111111111111111111111111"
        "00000000000000000000001000000000"
    };

Comment (vinniefalco) [inline include/beast/http/detail/basic_parser.hpp]: OOH!!! SHINY

Comment (vinniefalco) [inline include/beast/http/detail/basic_parser.hpp]: Can't initialize `constexpr` unfortunately but nice idea.

Comment (nbougalis) [inline include/beast/http/detail/basic_parser.hpp]: Or... you know, something like this:

```
inline
constexpr
bool
is_text(char c)
{
    auto const cc = static_cast<std::uint8_t>(c);

    if (cc < 32)
        return (cc == 9);

    return (cc != 127);
}
```

Comment (vinniefalco) [inline include/beast/http/detail/basic_parser.hpp]: That kills performance

Comment (nbougalis) [inline include/beast/http/detail/basic_parser.hpp]: Not as much as you'd think.

Comment (vinniefalco) [inline include/beast/http/detail/basic_parser.hpp]: You gotta understand though, its a cut-throat world out there and Beast is fighting for its life. There's a benchmark executable that compares the performance of Beast versus nodejs http-parser and even tiny differences matter:
https://travis-ci.org/vinniefalco/Beast/jobs/127369018#L448

**Pull request merged**